### ISSUE TITLE
Mastodon keeps trying to POST remote inbox with a public key already gone

### ISSUE THREAD

**Issue by ptrcnull** [mastodon/mastodon#33290]
### Steps to reproduce the problem

1. have a fediverse instance that federates with `mastodon.social`
2. observe a lot of failing incoming POST requests in the logs
3. notice that for every failing POST to `/users/<name>/inbox` there's a 404 from the remote instance for the public key used for said POST

### Expected behaviour

Mastodon should stop trying to deliver updates when the user no longer exists

### Actual behaviour

Mastodon keeps trying to send updates even though the user's profile is no longer available

### Detailed description

two examples from my instance's logs, with the instance other than `mastodon.social` censored, but still using Mastodon:
```
timestamp="2024-12-12T21:33:27.105+01:00" func=server.init.func1.Logger.13.1 level=ERROR latency="187.980614ms" userAgent="http.rb/5.1.1 (Mastodon/4.1.20; +https://mastodon.domain.tld/)" method=POST statusCode=500 path=/users/foo/inbox clientIP=205.166.94.38 pubKeyID=https://mastodon.domain.tld/users/bar#main-key errors="Error #01: callForPubKey: error dereferencing public key https://mastodon.domain.tld/users/bar#main-key: Dereference: GET request to https://mastodon.domain.tld/users/bar#main-key failed: status=\"404 Not Found\" body=\"{\"error\":\"Not Found\"}\"\n" requestID=gp6s7f4k04000va00d50 msg="Internal Server Error: wrote 54B"
timestamp="2024-12-12T21:36:02.819+01:00" func=server.init.func1.Logger.13.1 level=ERROR latency="84.574218ms" userAgent="Mastodon/4.4.0-alpha.1+pr-33266-17e66a4 (http.rb/5.2.0; +https://mastodon.social/)" method=POST statusCode=500 path=/users/foo/inbox clientIP=188.245.228.25 pubKeyID=https://mastodon.social/users/Guest17#main-key errors="Error #01: AuthenticateFederatedRequest: error dereferencing account https://mastodon.social/users/Guest17: enrichAccount: error webfingering remote account Guest17@mastodon.social: fingerRemoteAccount: error webfingering @Guest17@mastodon.social: account has been deleted/is gone\n" requestID=5vq9bf4k04000gvc5axg msg="Internal Server Error: wrote 54B"
```

### Mastodon instance

mastodon.social et al.

### Mastodon version

4.4.0-alpha.1+pr-33266-17e66a4

**Comment by trwnh** [comment]
i think the problem here is that 404 is "Not Found", which is not the same as 410 "Gone".

that said: i think that mastodon should probably consider 404 as well in addition to 410 for its liveness checks

**Comment by ptrcnull** [comment]
i don't think error code is the issue here; i'm getting the same behaviour (POSTs with the originating user being missing already) even when webfingering said user returns `410 Gone`.

**Comment by ClearlyClaire** [comment]
I assume what you are seeing is `Delete` activities that are sent from Mastodon to notify you of an account deletion.

Indeed, when a Mastodon account is deleted, its deletion is broadcasted to known servers to ensure proper deletion. The activity is a `Delete` activity with the `actor` and `object` being the same identifier. Such activities are signed with the actor's signing key, so that every server who already knows of the account can swiftly proceed to deleting it.

If you do not know about the actor already, you can safely ignore such activities and skip fetching the key.

**Comment by TheOneric** [comment]
Mastodon indeed replies with a 410 status for deleted actors afaict. Delivering actor `Delete`s to all known instances also isn’t unreasonable on its own.

However, Mastodon apparently also is quite eager to redeliver actor `Delete`s to instances after receiving a `401 Unauthorized` response. A `401` response indicates the signature could not be verified, most likely because the key was never known and no can no longer be fetched since the actor is already gone. There’s no point in reattempting delivery and doing so only wastes resources both on the originating Mastodon server and all its known remote instances; the latter of which will repeatedly try to fetch the gone key and/or actor.

It’s not reasonable to ask all other implementation to special case actor Deletes in the signature-verification-failed path, but Mastodon could easily and imho should just stop delivery attempts of actor `Delete`s on a `401` response. Here are some examples from `mastodon.social` actors:
```
# req-count  key_id   first_req   last_req
 18     https://mastodon.social/users/yos_final#main-key        2025-01-13 09:29:05.870     2025-01-13 11:18:10.467
 18     https://mastodon.social/users/yme_quispel#main-key      2025-01-13 19:32:27.372     2025-01-13 21:22:40.438
 18     https://mastodon.social/users/vero__85#main-key         2025-01-14 12:47:06.282     2025-01-14 14:29:58.820
 18     https://mastodon.social/users/u888b_org#main-key        2025-01-14 01:06:23.125     2025-01-14 02:45:10.537
 18     https://mastodon.social/users/torgiacat#main-key        2025-01-14 12:59:39.148     2025-01-14 14:46:20.207
 18     https://mastodon.social/users/tassos_c#main-key         2025-01-15 17:36:22.116     2025-01-15 19:27:04.133
 18     https://mastodon.social/users/stella_george12#main-key  2025-01-14 08:09:57.925     2025-01-14 09:58:58.238
```

**Comment by ClearlyClaire** [comment]
I think it *is* reasonable to ask other implementations to special-case what would be no-ops, however, deciding to retry on 401 on this specific scenario is definitely an oversight on Mastodon's side. I'll have a look, thanks for spotting this!

**Comment by ClearlyClaire** [comment]
I have opened https://github.com/mastodon/mastodon/pull/33617, but as described here, I have some reservation regarding other possible failure cases.

**Comment by naturzukunft** [comment]
> i don't think error code is the issue here; i'm getting the same behaviour (POSTs with the originating user being missing already) even when webfingering said user returns `410 Gone`.

Yes, same on my server. 410 GONE is not working.

**Comment by ClearlyClaire** [comment]
Do you have an example of such a deleted account that returns 410 Gone but is still retried by Mastodon? I can have a look and see if something's missing.

**Comment by naturzukunft** [comment]
> Do you have an example of such a deleted account that returns 410 Gone but is still retried by Mastodon? I can have a look and see if something's missing.

this is a very old endpoint, it hasn't existed for weeks:
`172.16.0.3 - - [20/Jan/2025:14:10:44 +0100] "POST /resource/acto0b7cf49c-5b74-413b-b9d1-960e76a4a7e3/inbox HTTP/1.1" 410`

Call to an existing endpoint but for an actorid, that is not available any more.
`172.16.0.3 - - [20/Jan/2025:14:10:37 +0100] "GET /resource/1bdbbb62-3bb1-4a9c-8f7d-4a56e1696f4e HTTP/1.1" 410`

I am quite sure that the calls come from mastodon.social. These are the urls for it.
- https://dev.rdf-pub.org/resource/acto0b7cf49c-5b74-413b-b9d1-960e76a4a7e3/inbox
- https://dev.rdf-pub.org/resource/1bdbbb62-3bb1-4a9c-8f7d-4a56e1696f4e

**Comment by ClearlyClaire** [comment]
Indeed, we do remove deleted accounts if we encounter 410 when refreshing their profile, but not in that delivery path.

Changing that will be a little bit of work, as by the time we perform the delivery we don't know which account we are delivering to (we collect inbox URIs beforehand and deduplicate them), so we cannot easily get back to the account to remove (the lack of database indexes makes this costly). This would be very worthwhile though.
